### PR TITLE
Blog App : Integration specs for views and Fix n+1 problems

### PR DESCRIPTION
**Pull request by naanmohammed**
In this milestone we:

- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user.
- Used Capybara to write integration tests for each view in your project.

BEFORE :
Screenshot (339)

AFTER:
Screenshot (340)

### CODE REVIEW

File: spec/users_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    expect(page).to have_content(@user.bio)
+  end
+  it 'shows the first 3 posts of the user' do
+    @user.recent_posts.each do |post|
+      expect(page).to have_content(post.title)
+      expect(page).to have_content(post.text)
+      expect(page).to have_content(post.comments_counter)
+      expect(page).to have_content(post.likes_counter)
+    end
+  end
```
Comment (danmainah) [inline spec/users_show_spec.rb]: - [ ] @naanmohammed and @amentetteh Great work on your tests for the user show page, however in this test, we were supposed to test if using the recent_posts method we can view only view user's first 3 posts, kindly consider modifying this test

File: spec/users_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  it "is redirected to that post's show page." do
+    click_link(@post1.title, href: "/users/#{@user.id}/posts/#{@post1.id}")
+    expect(page).to have_current_path(user_post_path(@post1.author, @post1))
+  end
+  it "redirects me to the user's posts index page on clicking to see all posts" do
+    click_link('See all posts')
+    expect(page).to have_current_path(user_posts_path(@user))
+  end
```
Comment (danmainah) [inline spec/users_show_spec.rb:66]: - [ ] @naanmohammed and @amentetteh Great work on your tests for the user show page, however, you forgot to add a test that a user `can see a button that lets the user view all of a user's posts.` in this case `See all posts`

File: spec/posts_index_spec.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+  it 'displays a post text on user posts page' do
+    visit user_posts_path(user_id: user.id)
+    expect(page).to have_content(post.text)
+  end
+
+  it 'displays the first comments on a post' do
+    visit user_posts_path(user_id: user.id)
+    expect(page).to have_content(comment.text)
+  end
```
Comment (danmainah) [inline spec/posts_index_spec.rb:44]: - [ ] @team a great job adding a test to check if the page displays the  first comments on a post, however, we are supposed to check if the `recent_comments` method is displaying only 5 recent comments

File: spec/posts_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  it 'displays the number of likes on a post' do
+    visit user_posts_path(user_id: user.id)
+    expect(page).to have_content(post.likes_counter)
+  end
+
+  it 'redirects to post show page when clicking on post title' do
+    visit user_posts_path(user_id: user.id)
+    click_link post.title
+    expect(page).to have_current_path(user_post_path(user_id: user.id, id: post.id))
+  end
```
Comment (danmainah) [inline spec/posts_index_spec.rb:60]: - [ ] @Team great job adding tests for the posting page, however, you forgot to add a test for the `that asserts the user can see the user's profile picture` and test `that asserts a user can see a section for pagination if there are more posts than fit on the view.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @naanmohammed and @amentetteh , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
✔️  Fixed the N + 1 problem
✔️  Added tests for all 4 pages and ensured all tests are passing


# Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

File: spec/posts_index_spec.rb (hunk 3 above)
Comment (naanmohammed) [reply spec/posts_index_spec.rb:44]: @danmainah Thank you for the review. The requirement explicitly asks to create a test for "I can see the first comments on a post". There is no requirement to check if the method is displaying 5 recent comments.

image
image

**Review by AshakaE — CHANGES_REQUESTED** [review]
Hi @naanmohammed 🙋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights


 ## Required Changes ♻️

- [ ] Kindly work on the changes requested by the previous reviewer 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something
**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by naanmohammed** [comment]
Hey @AshakaE. I forgot to push that was why. I had already made the changes.

File: spec/posts_index_spec.rb (hunk 3 above)
Comment (danmainah) [reply spec/posts_index_spec.rb:44]: @naanmohammed I agree, technically it doesn't show the user needs to check if the method displays 5 comments

File: spec/posts_index_spec.rb (hunk 4 above)
Comment (danmainah) [reply spec/posts_index_spec.rb:60]: - [ ] @naanmohammed and @amentetteh great work implementing all changes requested, however, forgot to do one requested change adding a test  `that asserts a user can see a section for pagination if there are more posts than fit on the view.`Kindly consider adding it.
image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @naanmohammed and @amentetteh , 

Great work on making the changes requested by a previous reviewer 👏  👏 👏
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
✔️ Fixed the N + 1 problem
✔️ Added tests for all 4 pages and ensured all tests are passing

 # Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by naanmohammed** [comment]
@danmainah We did all the changes you requested on your first review. It is not fair that you are introducing a new request now. We are being forced to use an extra review. All in all, Thank you for the review!

**Comment by danmainah** [comment]
> @danmainah We did all the changes you requested on your first review. It is not fair that you are introducing a new request now. We are being forced to use an extra review. All in all, Thank you for the review!

no I didn't introduce any new changes check my first review I requested the change while requesting for user profile picture, kindly check the screenshot below
image

**Review by TedLivist — APPROVED** [review]
## STATUS: APPROVED ✔️ ✔️

Hi @naanmohammed and @amentetteh,

✔️ Well-written and passing tests

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by naanmohammed** [comment]
Thank you @TedLivist

**Comment by prolajumokeoni** [comment]
Hi @amentetteh,

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted the wrong pull request, kindly create a new pull  in your own repository, please see the [Work and submission mode](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/projects/intergration_spec_and_n_1_project.md#work-and-submission-mode), please note that your partner has to commit to your own repository 👍🏿  

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification @prolajumokeoni.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Pull request merged**